pca.py

```diff
@@ -1,70 +1,178 @@
+            X -= new_mean
+
+        if self.N_:
+            if self.fixed_mean:
+                mean_cor = torch.zeros_like(self.mean_)
+            else:
+                mean_cor = torch.sqrt((self.N_ / n_total_samples) * n_samples) * (self.mean_ - new_mean)
+                self.mean_ = updated_mean
+
+            X = torch.vstack((self.singular_values_.unsqueeze(1) * self.components_, X, mean_cor))
+
+        U, S, Vt = torch.linalg.svd(X, full_matrices=False)
+        U, Vt = self._svd_flip(U, Vt, u_based_decision=False)
+
+        self.N_ = n_total_samples
+        self.components_ = Vt[:self.n_components]
+        self.singular_values_ = S[:self.n_components]
+
+        return self
 
     def transform(self, X):
-        assert hasattr(self, "components_"), "PCA must be fit before use."
-        return torch.matmul(X - self.mean_, self.components_.t())
+        return torch.mm(X - self.mean_, self.components_.T)
 
-    def fit_transform(self, X):
-        self.fit(X)
-        return self.transform(X)
+    def forward(self, X, check=True):
+        if self.training:
+            return self.fit(X, check)
+        else:
+            if self.N_:
+                return self.transform(X)
+            raise RuntimeError('PCA has not been fitted')
 
-    def inverse_transform(self, Y):
-        assert hasattr(self, "components_"), "PCA must be fit before use."
-        return torch.matmul(Y, self.components_) + self.mean_
 
-if __name__ == "__main__":
+if __name__ == '__main__':
     import numpy as np
-    from sklearn.decomposition import PCA as sklearn_PCA
-    from sklearn import datasets
-    iris = torch.tensor(datasets.load_iris().data)
-    _iris = iris.numpy()
-    devices = ['cpu']
-    if torch.cuda.is_available():
-        devices.append('cuda')
-    for device in devices:
-        iris = iris.to(device)
-        for n_components in (2, 4, None):
-            _pca = sklearn_PCA(n_components=n_components).fit(_iris)
-            _components = torch.tensor(_pca.components_)
-            pca = PCA(n_components=n_components).to(device).fit(iris)
-            components = pca.components_
-            assert torch.allclose(components, _components.to(device))
-            _t = torch.tensor(_pca.transform(_iris))
-            t = pca.transform(iris)
-            assert torch.allclose(t, _t.to(device))
-        __iris = pca.inverse_transform(t)
-        assert torch.allclose(__iris, iris)
-    print("passed!")
+    from sklearn.decomposition import IncrementalPCA as sklearn_IPCA
+    from sklearn.datasets import make_classification
+
+    n_components = 5
+    batch_size = 20
+
+    X, _ = make_classification(n_samples=100, n_features=20, random_state=0)
+    X = torch.tensor(X, dtype=torch.float32)
+
+    n_batches = X.shape[0] // batch_size
+
+    sklearn_ipca = sklearn_IPCA(n_components=n_components)
+
+    ipca = IncrementalPCA(n_components=n_components, n_features=X.shape[-1]).cuda().train()
```

Comment:
This assumes cuda is available, it would be better to iterate over available devices like in the previous version of the code.

Reply:
Improved, see my latest commit